crates/types/src/consensus.rs

```diff
@@ -882,100 +884,125 @@ impl<TYPES: NodeType> Consensus<TYPES> {
+            true,
+            |leaf, _, _| {
                 tracing::trace!(
                     "last_visited_view_number = {}, leaf.view_number = {}",
                     *last_visited_view_number,
                     *leaf.view_number()
                 );
 
-                if leaf.view_number() == qc_view {
+                if leaf.view_number() == leaf_view {
                     return true;
                 }
 
                 if last_visited_view_number - 1 != leaf.view_number() {
                     tracing::trace!("The chain is broken. Non consecutive views.");
-                    is_qc_extended = false;
+                    is_leaf_extended = false;
                     return false;
                 }
-                if high_qc_block_number != leaf.height() {
+                if leaf_block_number != leaf.height() {
                     tracing::trace!("The chain is broken. Block numbers do not match.");
-                    is_qc_extended = false;
+                    is_leaf_extended = false;
                     return false;
                 }
                 last_visited_view_number = leaf.view_number();
                 true
-            })
-        {
-            is_qc_extended = false;
+            },
+        ) {
+            is_leaf_extended = false;
             tracing::trace!("The chain is broken. Leaf ascension failed.");
             tracing::debug!("Leaf ascension failed; error={e}");
         }
-        tracing::trace!("Is the given QC an eQC? {}", is_qc_extended);
-        is_qc_extended
+        tracing::trace!("Can the given leaf form an eQC? {}", is_leaf_extended);
+        is_leaf_extended
     }
 
-    /// Return true if the given Quorum Certificate takes part in forming an eQC, i.e.
-    /// it is one of the 3-chain certificates but not the eQC itself
-    pub fn is_qc_forming_eqc(&self, cert: &QuorumCertificate<TYPES>) -> bool {
-        self.is_qc_for_last_block(cert) && !self.is_qc_extended(cert)
+    /// Returns true if a given leaf is for the last block in the epoch
+    pub fn is_leaf_for_last_block(&self, leaf_commit: LeafCommitment<TYPES>) -> bool {
+        let Some(leaf) = self.saved_leaves.get(&leaf_commit) else {
+            return false;
+        };
+        let block_height = leaf.height();
+        if block_height == 0 || self.epoch_height == 0 {
+            false
+        } else {
+            block_height % self.epoch_height == 0
+        }
     }
 
-    /// Return true if the high QC takes part in forming an eQC, i.e.
-    /// it is one of the 3-chain certificates but not the eQC itself
-    pub fn is_high_qc_forming_eqc(&self) -> bool {
-        self.is_high_qc_for_last_block() && !self.is_high_qc_extended()
+    /// Returns an `Epoch` for the view following the given view.
+    /// There are three cases:
+    /// 1. The leaf from the given view is not for the last block.
+    ///    The epoch should be calculated based on the next block number.
```

Comment:
I don't think I understand this case, if the leaf is not the last block (say it's height 3 and epoch length is 100), then isn't this the trivial case and can just be handled without going into the validated state map?  We already got the height before calling this function right?  

Basically I expect the code on forming eqc to just do the decide rule, and if the block decided is the last of the epoch we can advance, all other cases we stay in the current epoch.  I can see how the logic kind of does that, but it's hard to follow when the case for "I'm not even trying to forma an eqc" and "I just formed an eqc" end up being handled by the same line of code and It's a little confusing that `is_leaf_forming_eqc` is false when we form the eqc?